src/components/grid-layer/grid-layer-component.jsx

```diff
@@ -31,74 +39,57 @@ const GridLayer = ({map, view, setGridCellData, setGridCellGeometry, handleGlobe
-      })
+    loadModules(['esri/layers/FeatureLayer']).then(([FeatureLayer]) => {
+      const layer = new FeatureLayer({
+        url: BIODIVERSITY_FACETS_SERVICE_URL
+      });
+      setBiodiversityFacetsLayer(layer);
+    })
   }, [])
 
   // set the view extent when view stationary
   useEffect(() => {
-      watchHandle = watchUtils && watchUtils.whenTrue(view, "stationary", function() {
-        setViewExtent(view.extent);
-      })
+    watchHandle = watchUtils && watchUtils.whenTrue(view, "stationary", function() {
+      setViewExtent(view.extent);
+    })
     return function cleanUp() {
       watchHandle && watchHandle.remove();
     }
   },[watchUtils])
 
-   // update gridcells when view extent changes
-   useEffect(() => {
-        const { extent } = view;
-        const scaledDownExtent = extent.clone().expand(0.9);
-        handleGlobeUpdating(true);
-        watchUpdateHandle = gridViewLayer && gridViewLayer.watch('updating', function(value) {
-          if (!value) {
-            queryHandle && (!queryHandle.isFulfilled()) && queryHandle.cancel();
-            queryHandle = gridViewLayer.queryFeatures({
-              geometry: extent,
-              spatialRelationship: 'intersects',
-              returnGeometry: true
-            }).then(function(results) {
-              const containedGridCells = results.features.filter(gridCell => scaledDownExtent.contains(gridCell.geometry.extent));
-              const hasContainedGridCells = containedGridCells.length > 0;
-              const singleGridCell = results.features.filter(gridCell => gridCell.geometry.contains(view.center));
-              // If there are not a group of cells pick the one in the center
-              const gridCells = hasContainedGridCells ? containedGridCells : singleGridCell;
-              // Change data on the store and paint only when grid cell chaged
-              handleGlobeUpdating(false);
-              if (!cellsEquality(gridCellRef.current, gridCells, hasContainedGridCells)) {
-                // dispatch action
-                setGridCellData(gridCells.map(c => c.attributes));
-                loadModules(["esri/geometry/geometryEngine"])
-                .then(([geometryEngine]) => {
-                  // create aggregated grid cell geometry
-                  const gridCellGeometry = calculateAgregatedGridCellGeometry(hasContainedGridCells, gridCells, geometryEngine);
-                  // paint it
-                  if (gridCellGraphic) { gridCellGraphic.geometry = gridCellGeometry };
-                  // Add it to the store
-                    setGridCellGeometry(gridCellGeometry)
-                  })
-              }
-              gridCellRef.current = gridCells;
-            })
+  useEffect(() => {
+    if (biodiversityFacetsLayer && gridCellGraphic) {
+      const containedCellsQueryObject = containedQuery(biodiversityFacetsLayer, view.extent);
+      biodiversityFacetsLayer.queryFeatures(containedCellsQueryObject)
```

Comment:
I think we can extract this bit to a function passing `query` and type of a cell; `singleCell` or `aggregatedCells` as it is almost the same in the if and else body

Reply:
I've extracted the internal logic after the `query` on the `if/else`  statement to a function. But I'm afraid that the query needs to seat outside of this function as they are nested queries, if we do not get multiple grid cells only then we perform a second query inside the `callback`

Reply:
thank you 💯